Added avoid repeating module name rule

corroded opened the pull request:
This is for #102

eksperimental commented on README.md:
```diff
@@ -774,6 +774,32 @@ Translations of the guide are available in the following languages:
 
+* <a name="avoid-repeating-aliased-module-names"></a>
+  If you plan to alias your modules, avoid using the same name consecutively.
+  This avoids any problems with [conflicting aliases](https://elixirforum.com/t/using-aliases-for-fubar-fubar-named-module/1723/3)
```
I would add a link to the beginning of the thread. not to the middle.

corroded replied:
Sorry, pasted the wrong link lol. All good 👍

eksperimental commented on README.md:
````diff
@@ -774,6 +774,32 @@ Translations of the guide are available in the following languages:
+
+  ```elixir
+  # not preferred
+
````
:scissors:

eksperimental commented on README.md:
```diff
@@ -774,6 +774,32 @@ Translations of the guide are available in the following languages:
+  end
+
+  # somewhat preferred
+
```
:scissors:

eksperimental commented on README.md:
```diff
@@ -774,6 +774,32 @@ Translations of the guide are available in the following languages:
+  end
+
+  # preferred
+
```
:scissors:

corroded commented:
@eksperimental addressed your comments :) Let me know if it needs more work. For examples, I just copied it from the original issue - I am still not that adept in Elixir to provide a better example (this is my way of studying/learning it)

eksperimental commented on README.md:
````diff
@@ -774,6 +774,29 @@ Translations of the guide are available in the following languages:
   ```
 
+* <a name="avoid-repeating-aliased-module-names"></a>
+  If you plan to alias your modules, avoid using the same name consecutively.
````
I would replace "name" with "fragment"

eksperimental commented on README.md:
```diff
@@ -774,6 +774,29 @@ Translations of the guide are available in the following languages:
+    alias Todo.Todo
+  end
+
+  # somewhat preferred
```
this is still not preferred. as you do not solve anything.

corroded replied:
Can you please help me with a better example for this? I was just following the original thread and the issue so not sure what a good example would be. To be honest, I would never name something consecutively like the example so really couldn't think of a use case where that is even possible.

eksperimental commented on README.md:
````diff
@@ -774,6 +774,29 @@ Translations of the guide are available in the following languages:
+    alias Todo.List
+  end
+  ```
+
````
I don't think this is clearly explained.
The whole reason for not repeating fragments in the module name is because when they are aliased, it could be ambiguous to which module you are refereing to.

examples.
```elixir
# not preferred
defmodule Todo.Todo do
  ...
end

defmodule AnotherModule do
  alias Todo.Todo
  # here it is ambigious to which one we refer when we call `Todo`
end

I think those would be more real life examples, as in the examples given the proper way to refer to module from within is using `__MODULE__` not matter what is called. the problem is when aliasing from another module.

corroded commented:
@DavidAntaramian @eksperimental how about this:

```elixir
  # not preferred
  defmodule Todo.Todo do
    ...
  end

  defmodule AnotherModule do
    alias Todo.Todo
    # here it is ambigious to which one we refer when we call `Todo`
  end

  # preferred
  defmodule Todo.List do
    ...
    # Basically, you should consider a better name for your module
    # if you ever find yourself duplicating fragments.
  end
```

I think if you're duplicating your module names/fragments, it is already a code smell. I totally agree with both your sentiments (including Jose's) that you should just pick a better name than duplicating something. It is too ambiguous and confusing.

christopheradams commented:
I think it can be simplified. I think it just needs two brief examples with no additional comments of the style that is not preferred (`Todo.Todo`) and preferred (`Todo.Item` or `Todo.Note`). The reasons are clearer naming structure and no chance of ambiguous or confusing aliases.

christopheradams commented:
* Don't repeat fragments in module names and namespaces. This improves overall readability and avoids [ambiguous aliases][Conflicting Aliases].

```elixir
# not preferred
defmodule Todo.Todo do
  ...
end

# preferred
defmodule Todo.Item do
  ...
end
```

[Conflicting Aliases]: https://elixirforum.com/t/using-aliases-for-fubar-fubar-named-module/1723

corroded commented:
@christopheradams I've changed this to what you have said 👍  thanks!
